core/field_number.js

```diff
@@ -32,23 +32,45 @@ goog.require('goog.userAgent');
 
 
 /**
+ * Return an appropriate restrictor, depending on whether this FieldNumber
+ * allows decimal or negative numbers.
+ * @param {boolean} decimalAllowed Whether number may have decimal/float component.
+ * @param {boolean} negativeAllowed Whether number may be negative.
+ * @return {!RegExp} Regular expression for this FieldNumber's restrictor.
+ */
+var getNumRestrictor = function(decimalAllowed, negativeAllowed) {
+  var pattern = "[\\d]|[e]"; // Always allow digits and e, from Scratch 2.0.
+  if (decimalAllowed) {
+    pattern += "|[\\.]";
+  }
+  if (negativeAllowed) {
+    pattern += "|[-]";
+  }
+  return new RegExp(pattern);
+};
+
+/**
  * Class for an editable number field.
+ * In scratch-blocks, the min/max/precision properties are only used
+ * to construct a restrictor on typable characters, and to inform the pop-up numpad on touch devices.
+ * These properties are included here (i.e. instead of just accepting a decimalAllowed, negativeAllowed)
+ * to maintain API compatibility with Blockly and Blockly for Android.
  * @param {string} text The initial content of the field.
+ * @param {number} min Minimum number allowed. NOTE: only used to determine if negatives are allowed.
+ * @param {number} max Maximum number allowed. NOTE: unused in scratch-blocks.
+ * @param {number} precision Step allowed between numbers. NOTE: only used to determine if decimal values are allowed.
  * @param {Function=} opt_validator An optional function that is called
  *     to validate any constraints on what the user entered.  Takes the new
  *     text as an argument and returns the accepted text or null to abort
  *     the change.
- * @param {number} precision Precision of the decimal value (negative power of 10).
- * @param {number} min Minimum value of the number.
- * @param {number} max Maximum value of the number.
  * @extends {Blockly.FieldTextInput}
  * @constructor
  */
-Blockly.FieldNumber = function(text, opt_validator, precision, min, max) {
-  this.precision_ = precision;
-  this.min_ = min;
-  this.max_ = max;
-  Blockly.FieldNumber.superClass_.constructor.call(this, text, opt_validator);
+Blockly.FieldNumber = function(text, opt_validator, min, max, precision) {
+  this.decimalAllowed_ = precision < 1;
```

Comment:
You should use floor(precision) == precision to check for decimals. 1.2 is still a valid precision.